### PR TITLE
add lua api: script.execute_as_script

### PR DESCRIPTION
**Pull request by Alice2333g**

### CODE REVIEW

File: src/lua/bindings/script.cpp [hunk 1]
```diff
@@ -151,15 +151,26 @@ namespace lua::script
 		module->m_registered_scripts.push_back(std::move(lua_script));
 	}
 
+	// Lua API: function
+	// Table: script
+	// Name: execute_as_script
+	// Param: script_name: string: target script thread.
+	// Param: func: function: function that will be executed once in the script thread.
+	static void execute_as_script(const std::string& script_name, sol::function func)
```
Comment (xiaoxiao921) [inline src/lua/bindings/script.cpp]: to keep it consistent with the rest of the bindings use `sol::protected_function` for the `func` param

**Review by xiaoxiao921 — CHANGES_REQUESTED** [review]

**Comment by xiaoxiao921** [comment]
Looks like it doesnt even build, was it even tested properly?

**Comment by Alice2333g** [comment]
> 看起来它甚至没有构建，它甚至经过了适当的测试吗？

working for me.

**Comment by gir489returns** [comment]
@xiaoxiao921 It literally failed for no reason.

`[630/781] Building CXX object CMakeFiles\YimMenu.dir\src\lua\lua_module.cpp.obj
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 1.`

**Comment by Alice2333g** [comment]
> @xiaoxiao921 It literally failed for no reason.
> 
> `[630/781] Building CXX object CMakeFiles\YimMenu.dir\src\lua\lua_module.cpp.obj ninja: build stopped: subcommand failed. Error: Process completed with exit code 1.`

It was successfully generated on my PC

edit: Okay, I forgot to add ``` #include "gta_util.hpp" ```

**Comment by xiaoxiao921** [comment]
> @xiaoxiao921 It literally failed for no reason.
> 
> `[630/781] Building CXX object CMakeFiles\YimMenu.dir\src\lua\lua_module.cpp.obj ninja: build stopped: subcommand failed. Error: Process completed with exit code 1.`

It always fail for a reason. Though not always obvious :p

File: src/lua/bindings/script.cpp [hunk 2]
```diff
@@ -156,9 +157,9 @@ namespace lua::script
 	// Name: execute_as_script
 	// Param: script_name: string: target script thread.
 	// Param: func: function: function that will be executed once in the script thread.
-	static void execute_as_script(const std::string& script_name, sol::function func)
+	static void execute_as_script(const std::string& script_name, sol::protected_function func_, sol::this_state state)
```
Comment (gir489returns) [inline src/lua/bindings/script.cpp:160]: Change from func_ to something reasonable, like parameter_func or something.

**Review by gir489returns — CHANGES_REQUESTED** [review]

File: src/lua/bindings/script.cpp (hunk 2 above)
Comment (Alice2333g) [reply src/lua/bindings/script.cpp]: example code:
```
function run()
	log.info(SCRIPT.GET_THIS_SCRIPT_NAME())
end

script.run_in_fiber(function(state)
	run() -- main_persistent
	script.execute_as_script("shop_controller", function()
		run() -- shop_controller
	end)
	log.info("function")
	script.execute_as_script("freemode", run) -- freemode
	log.info("function")
end)
```

Comment (gir489returns) [reply src/lua/bindings/script.cpp:160]: What does that have to do with what I asked?

Comment (Alice2333g) [reply src/lua/bindings/script.cpp]: in some natives, it needs to be used here.

Comment (gir489returns) [reply src/lua/bindings/script.cpp:160]: I asked you to change the name of the parameter.

**Review by gir489returns — DISMISSED** [review]

**Comment by gir489returns** [comment]
@xiaoxiao921 This is still pending. Any particular reason why?

**Review by xiaoxiao921 — APPROVED** [review]

**Pull request merged**